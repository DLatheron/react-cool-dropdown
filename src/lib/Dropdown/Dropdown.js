import React, { useMemo, useRef, useReducer } from 'react';
import { useOutsideClick } from '../Hooks';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCaretUp, faCaretDown, faTimes, faBreadSlice } from '@fortawesome/free-solid-svg-icons'

import ListContainer from './List/ListContainer';
import SelectionContainer from './Selection/SelectionContainer';

import './Dropdown.scss';

function initialState({ open }) {
    return {
        open,
        selected: [],
        searchTerm: ''
    };
}

function reducer(state, action) {
    switch (action.type) {
        case 'open':
            return {
                ...state,
                open: true
            };

        case 'close':
            return {
                ...state,
                open: false,
                searchTerm: ''
            };

        case 'toggle':
            return {
                ...state,
                open: !state.open
            };

        case 'clearSearch':
            return {
                ...state,
                searchTerm: ''
            };

        case 'setSearchTerm':
            return {
                ...state,
                searchTerm: action.searchTerm
            };

        case 'clearSelected':
            return {
                ...state,
                selected: []
            };

        case 'setSelected':
            return {
                ...state,
                selected: action.selected
            };

        default:
            throw new Error(`Uncognised action type ${action.type}`);
    }
}

Dropdown.defaultProps = {
    open: false,
    prefix: undefined,
    suffix: undefined,
    options: [],
    sortFn: undefined,
    filterFn: undefined,
    maxSelected: 1,
    clear: <FontAwesomeIcon icon={faTimes} />,
    handle: ({ state }) => <FontAwesomeIcon icon={state.open ? faCaretUp : faCaretDown} />,
};

export default function Dropdown(props) {
    const dropdownRef = useRef();
    const searchRef = useRef();

    const [state, dispatch] = useReducer(reducer, initialState({ open: props.open }));

    const sortedOptions = useMemo(() => (
        props.sortFn
            ? [...props.options].sort(props.sortFn)
            : props.options
    ), [props.options, props.sortFn]);

    const {
        filterFn
    } = props;

    const filteredOptions = useMemo(() => (
        (filterFn && state.searchTerm !== '')
            ? sortedOptions.filter(item => filterFn(state.searchTerm, item))
            : sortedOptions
    ), [sortedOptions, state.searchTerm, filterFn]);

    // TODO: If maxSelected is not 0 and the state.selected.length is greater than the maxSelected then chop off the end of the selection...

    const methods = {
        toggleDropdown: () => {
            dispatch({ type: 'toggle' });
        },
        closeDropdown: () => {
            dispatch({ type: 'close' });
        },
        openDropdown: () => {
            dispatch({ type: 'open' });
        },
        suppressEvent: event => {
            event.preventDefault();
            event.stopPropagation();
            event.nativeEvent.stopImmediatePropagation();
        },
        isSelected: item => state.selected.find(i => item === i),
        numSelected: () => state.selected.length,
        selectOption: item => {
            let newSelected;
            let open = state.open;
            let maxSelected = props.maxSelected;

            switch (maxSelected) {
                case 0:
                    if (methods.isSelected(item)) {
                        newSelected = [...state.selected.filter(i => item !== i)];
                    } else {
                        newSelected = [
                            ...state.selected,
                            item
                        ];
                    }
                    break;

                case 1:
                    newSelected = [item];
                    open = false;
                    break;

                default:
                    throw new Error(`TODO: Support a limited number of selections... ${maxSelected}`);
            }

            dispatch({ type: 'setSelected', selected: newSelected });
            if (open === false) {
                dispatch({ type: 'close' });
            }

            // Return focus to search.
            searchRef.current && searchRef.current.focus();
            // searchRef.current.scrollIntoView();
        },
        clearOption: item => {
            dispatch({ type: 'setSelected', selected: [...state.selected.filter(i => item !== i)] });
        },
        clearAll: () => {
            dispatch({ type: 'clearSelected' });
        },
        search: searchTerm => {
            dispatch({ type: 'setSearchTerm', searchTerm });
        }
    };

    // TODO: Need to make methods safe so they don't change...
    useOutsideClick(dropdownRef, methods.closeDropdown, state.open);

    const renderProps = {
        props,
        methods,
        state: {
            ...state,
            filteredOptions,
            searchRef
        }
    };

    return (
        <div
            ref={dropdownRef}
            className='dropdown'
            onClick={event => {
                if (!state.open) {
                    methods.suppressEvent(event);
                    methods.openDropdown();
                }
            }}
        >
            <SelectionContainer {...renderProps} />
            <ListContainer {...renderProps} />
        </div>
    );
}
