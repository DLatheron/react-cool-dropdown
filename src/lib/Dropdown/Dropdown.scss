.dropdown {
    >.selection-box {
        display: grid;
        grid-template-columns: auto 1fr auto auto auto;
        grid-template-areas: 'prefix selected clear suffix handle';
        align-items: stretch;

        height: 30px;
        border: 1px solid black;
        background-color: white;

        cursor: default;

        >.prefix {
            grid-area: prefix;
            margin: auto;
            padding: 0 10px;
            font-weight: bold;
        }

        >.selected {
            grid-area: selected;
            margin: auto 0;
            padding: 0 10px;

            overflow: auto;

            &::-webkit-scrollbar {
                display: none;
            }

            display: flex;
            flex: 1 1;

            .no-selected-option {
                width: fit-content;
                margin: 2px;
                padding: 2px 10px 2px 0;
                font-style: italic;
            }

            .single-selected-option {
                width: fit-content;
                text-overflow: ellipsis;
                white-space: nowrap;
                margin: 2px;
                padding: 2px 0;
            }

            .multi-selected-option {
                width: fit-content;
                text-overflow: ellipsis;
                white-space: nowrap;
                margin: 2px;
                padding: 2px 0 2px 10px;
                background-color: #e8e8e8;

                .icon {
                    font-size: 0.75em;
                    margin: auto 0;
                    padding: 0 10px;
                    cursor: pointer;
                }
            }

            .search {
                outline: none;
                border: none;
                width: fit-content;
                min-width: 200px;

                &::placeholder {
                    color: lightgrey;
                }
            }
        }

        >.clear {
            grid-area: clear;
            margin: auto;
            line-height: 30px;
            padding: 0 10px;

            cursor: pointer;
        }

        >.suffix {
            grid-area: suffix;
            margin: auto;
            padding: 0 10px;
            width: fit-content;
            height: fit-content;
        }

        >.handle {
            grid-area: handle;
            margin: auto;
            line-height: 30px;
            padding: 0 10px;

            cursor: pointer;
        }
    }

    >.anchor {
        position: relative;

        >.option-menu {
            position: absolute;
            top: 5px;
            left: 0;
            width: calc(100% - 2px);
            border: 1px solid red;
            background-color: white;

            &.open {
                display: block;
            }

            &.closed {
                display: none;
            }

            display: grid;
            grid-template-rows: auto 1fr auto;
            grid-template-areas: 'header' 'body' 'footer';

            >.header {
                grid-area: header;
                background-color: beige;
                padding: 10px;
                border-bottom: 1px solid black;
            }

            >.body {
                grid-area: body;
                max-height: 200px;
                overflow: auto;

                .option {
                    cursor: pointer;

                    padding: 0 10px;
                    margin: auto 0;
                    line-height: 32px;
                    height: 32px;

                    &:hover {
                        background-color: lightgray;
                    }

                    &.selected {
                        color: red;
                    }

                    &.none {
                        font-style: italic;
                    }
                }


                .no-options {
                    font-style: italic;
                    padding: 0 10px;
                    line-height: 32px;
                    color: lightgray;
                }
            }

            >.footer {
                grid-area: footer;
                background-color: beige;
                padding: 10px;
                border-top: 1px solid black;
            }
        }
    }
}