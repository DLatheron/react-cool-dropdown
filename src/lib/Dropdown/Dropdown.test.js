import React from 'react';
import Dropdown from './Dropdown';
import renderer from 'react-test-renderer';

describe('Dropdown', () => {
    it('renders properly', () => {
        const tree = renderer
            .create(<Dropdown label='Email' placeholder='name@example.com' />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
