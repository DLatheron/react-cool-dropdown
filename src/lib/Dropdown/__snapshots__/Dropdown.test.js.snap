// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders properly 1`] = `
<div
  className="simple-form-group"
>
  <label
    className="simple-text-label"
  >
    Email
  </label>
  <input
    className="simple-text-input"
    onChange={[Function]}
    placeholder="name@example.com"
    type="text"
    value={undefined}
  />
</div>
`;
